README.md

```diff
@@ -11,572 +11,528 @@
+Get the line, word, char counts in one chunk.
 
 ``` haskell
-import qualified Streamly.Prelude as S
-import Control.Concurrent
+import qualified Streamly.Data.Array.Foreign as Array
 
-delay n = S.yieldM $ do
-    threadDelay (n * 1000000)
-    tid <- myThreadId
-    putStrLn (show tid ++ ": Delay " ++ show n)
+countArray :: Array Word8 -> IO Counts
+countArray arr =
+      Stream.unfold Array.read arr            -- SerialT IO Word8
+    & Stream.decodeLatin1                     -- SerialT IO Char
+    & Stream.foldl' count (Counts 0 0 0 True) -- IO Counts
 ```
-### Serial
+When combining the counts in two contiguous chunks, we would also need to
+know whether the first element of the next chunk was a space char or
+non-space to know whether the same word is continuing to the next chunk or
+if it is a new word. So add that too, giving (firstCharWasSpace, Counts).
 
 ``` haskell
-main = S.drain $ delay 3 <> delay 2 <> delay 1
-```
+partialCounts :: Array Word8 -> IO (Bool, Counts)
+partialCounts arr = do
+    let r = Array.getIndex arr 0
+    case r of
+        Just x -> do
+            counts <- countArray arr
+            return (isSpace (chr (fromIntegral x)), counts)
+        Nothing -> return (False, Counts 0 0 0 True)
 ```
-ThreadId 36: Delay 3
-ThreadId 36: Delay 2
-ThreadId 36: Delay 1
-```
-
-### Parallel
 
+Combine the counts from two consecutive chunks.
 ``` haskell
-main = S.drain . S.fromParallel $ delay 3 <> delay 2 <> delay 1
-```
+addCounts :: (Bool, Counts) -> (Bool, Counts) -> (Bool, Counts)
+addCounts (sp1, Counts l1 w1 c1 ws1) (sp2, Counts l2 w2 c2 ws2) =
+    let wcount =
+            if not ws1 && not sp2 -- no space between two chunks
+            then w1 + w2 - 1
+            else w1 + w2
+     in (sp1, Counts (l1 + l2) wcount (c1 + c2) ws2)
 ```
-ThreadId 42: Delay 1
-ThreadId 41: Delay 2
-ThreadId 40: Delay 3
-```
-
-## Nested Loops (aka List Transformer)
-
-The monad instance composes like a list monad.
 
+Now put it all together, we only need to divide the stream into arrays,
+apply our counting function to each array and then combine all the counts.
 ``` haskell
-import qualified Streamly.Prelude as S
+wc :: String -> IO (Bool, Counts)
+wc file = do
+      Stream.unfold File.readChunks file -- AheadT IO (Array Word8)
+    & Stream.mapM partialCounts          -- AheadT IO (Bool, Counts)
+    & Stream.maxThreads numCapabilities  -- AheadT IO (Bool, Counts)
+    & Stream.fromAhead                   -- SerialT IO (Bool, Counts)
+    & Stream.foldl' addCounts (False, Counts 0 0 0 True) -- IO (Bool, Counts)
+```
 
-loops = do
-    x <- S.fromFoldable [1,2]
-    y <- S.fromFoldable [3,4]
-    S.yieldM $ putStrLn $ show (x, y)
+Note that `Stream.aheadly` is the only difference in a concurrent and
+non-concurrent program. If we remove that we still have a perfectly valid,
+well performing serial program. Notice, how succinctly and idiomatically
+we expressed the concurrent word counting problem.
```

Comment:
This is a better way to state observations than to have a separate heading for it.